### PR TITLE
Create view models `SitterMainTabViewModel` and `OwnerMainTabViewModel` for MainTabViews for sitter and owner.

### LINKED ISSUE #102: Create ViewModels for MainTabViews for sitter and owner.
## Depends on tasks:
- #_issue
- #_issue

### Dev requirements and restrictions:

- Create ViewModels for MainTabViews for sitter.
- Create ViewModels for MainTabViews for owner.
- Add method to change to nil for selectedRole in PreferencesStorage KeyValueStorage.

### PR DESCRIPTION
**Pull request by julyakmita**
## Summary
This PR is a part of the work to resolve #102
The main goal of the PR is to add method as a wrapper of method set() in PreferencesHandler to change selected role to `none` for `logout button`.

## Changes

- Add the SitterMainTabViewModel.
- Add the OwnerMainTabViewModel.
- Add tests for new models.

## Checks to complete
- [x] I've built and run my changes, and no warnings or errors occurred.
- [x] All existing tests passed with my changes.
- [x] My code follows our style guide.
- [x] I've performed a self-review of my code.
- [x] PR's title has a conventional commit style (short and descriptive).
- [x] Every non-private interface is documented properly.
- [x] I've added tests for my code (if it makes sense).
- [x] PR has assignee, reviewers, milestone, it is properly linked to the project and to the issue.

### CODE REVIEW

File: Woof/Woof/ViewModel/OwnerMainTabViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import Foundation
+
+/// The view model for a owner main tab view, responsible for preparing and processing data for the view.
+final class OwnerMainTabViewModel: ObservableObject {
+    /// Resets the current owner role  in the app by default value `none`.
+    func resertCurrentOwnerRole() {
```
Comment (iBYarchitect) [inline Woof/Woof/ViewModel/OwnerMainTabViewModel.swift:6]: typo:
```suggestion
    func resetCurrentOwnerRole() {
```

File: Woof/Woof/ViewModel/OwnerMainTabViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import Foundation
+
+/// The view model for a owner main tab view, responsible for preparing and processing data for the view.
+final class OwnerMainTabViewModel: ObservableObject {
+    /// Resets the current owner role  in the app by default value `none`.
```
Comment (iBYarchitect) [inline Woof/Woof/ViewModel/OwnerMainTabViewModel.swift:5]: typo
```suggestion
    /// Resets the current owner role in the app by default value `none`.
```

File: Woof/Woof/ViewModel/OwnerMainTabViewModel.swift (hunk 1 above)
Comment (dersim-davaod) [inline Woof/Woof/ViewModel/OwnerMainTabViewModel.swift:6]:
```suggestion
    func resetsCurrentOwnerRole() {
```

?

Could you please enable spell checking and/or install external spell checker?

File: Woof/Woof/ViewModel/OwnerMainTabViewModel.swift (hunk 2 above)
Comment (dersim-davaod) [inline Woof/Woof/ViewModel/OwnerMainTabViewModel.swift:5]:
```suggestion
    /// Resets the current owner role in the app by default value `none`.
```

File: Woof/Woof/ViewModel/OwnerMainTabViewModel.swift (hunk 1 above)
Comment (dersim-davaod) [inline Woof/Woof/ViewModel/OwnerMainTabViewModel.swift:6]:
```suggestion
    func resetCurrentRole() {
```

?

File: Woof/Woof/ViewModel/SitterMainTabViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import Foundation
+
+/// The view model for a pet sitter main tab view, responsible for preparing and processing data for the view.
+final class SitterMainTabViewModel: ObservableObject {
+    /// Resets the current sitter role  in the app by default value `none`.
+    func resertCurrentSitterRole() {
```
Comment (iBYarchitect) [inline Woof/Woof/ViewModel/SitterMainTabViewModel.swift:6]: typo:
```suggestion
    func resetCurrentSitterRole() {
```

File: Woof/Woof/ViewModel/SitterMainTabViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import Foundation
+
+/// The view model for a pet sitter main tab view, responsible for preparing and processing data for the view.
+final class SitterMainTabViewModel: ObservableObject {
+    /// Resets the current sitter role  in the app by default value `none`.
```
Comment (iBYarchitect) [inline Woof/Woof/ViewModel/SitterMainTabViewModel.swift:5]:
```suggestion
    /// Resets the current sitter role in the app by default value `none`.
```

File: Woof/Woof/ViewModel/OwnerMainTabViewModel.swift (hunk 1 above)
Comment (dersim-davaod) [reply Woof/Woof/ViewModel/OwnerMainTabViewModel.swift:6]: Also we use imperative verbs in the names of method.
Please read Swift API guideline carefully.

File: Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import XCTest
```
Comment (dersim-davaod) [inline Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift:1]: Could you please explain why these files are within the `OwnerProfileViewModelTests`?

File: Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+final class SitterMainTabViewModelTests: XCTestCase {
+    func testSitterMainTabViewModelExistsInAPI() {
+        SitterMainTabViewModel()
```
Comment (dersim-davaod) [inline Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift:5]: Do we need this test, taking into account we have the `testResertCurrentSitterRoleMethodExistsInAPI` ?

File: Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/OwnerMainTabViewModelTests.swift [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import XCTest
+
+final class OwnerMainTabViewModelTests: XCTestCase {
+    func testOwnerMainTabViewModelExistsInAPI() {
+        OwnerMainTabViewModel()
```
Comment (iBYarchitect) [inline Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/OwnerMainTabViewModelTests.swift:5]: try to build an app and look for warning messages, you should solve them
```suggestion
        _ = OwnerMainTabViewModel()
```

File: Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    }
+
+    func testResertCurrentSitterRoleMethodExistsInAPI() {
+        SitterMainTabViewModel().resertCurrentSitterRole()
+    }
+
+    func testResertCurrentSitterRoleMethod() {
```
Comment (dersim-davaod) [inline Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift:12]: The name of test does not show its intention.

**Review by dersim-davaod — CHANGES_REQUESTED** [review]
From the PR's description it's hard to understand why we need to implement this particular methods and how we're going to integrate them into the app.

The PR has the `Create view models for main tab views` name.
I know there're 3 tabs in the app, but I see only 2 new view models. Is that ok?

File: Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/OwnerMainTabViewModelTests.swift [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import XCTest
+
+final class OwnerMainTabViewModelTests: XCTestCase {
+    func testOwnerMainTabViewModelExistsInAPI() {
```
Comment (iBYarchitect) [inline Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/OwnerMainTabViewModelTests.swift:4]: seems to be clearer
```suggestion
    func testOwnerMainTabViewModelCanBeInitialized() {
```

File: Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/OwnerMainTabViewModelTests.swift [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+        OwnerMainTabViewModel()
+    }
+
+    func testResertCurrentOwnerRoleMethodExistsInAPI() {
+        OwnerMainTabViewModel().resertCurrentOwnerRole()
+    }
+
+    func testResertCurrentOwnerRoleMethod() {
```
Comment (iBYarchitect) [inline Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/OwnerMainTabViewModelTests.swift:12]: Please add comments
``` swift
//Given //When //Then
```

**Review by iBYarchitect — CHANGES_REQUESTED** [review]
Good job!
Please self-review the code for typos and corrections in the tests.

File: Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/OwnerMainTabViewModelTests.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (julyakmita) [reply Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/OwnerMainTabViewModelTests.swift:12]: Thanks for the suggestion! Fixed it.

File: Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/OwnerMainTabViewModelTests.swift (hunk 7 above)
Comment (julyakmita) [reply Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/OwnerMainTabViewModelTests.swift:5]: Thanks, sorry for this.
Fixed.

File: Woof/Woof/ViewModel/SitterMainTabViewModel.swift (hunk 4 above)
Comment (julyakmita) [reply Woof/Woof/ViewModel/SitterMainTabViewModel.swift:5]: Thanks for your attention! Fixed it.

File: Woof/Woof/ViewModel/OwnerMainTabViewModel.swift (hunk 2 above)
Comment (julyakmita) [reply Woof/Woof/ViewModel/OwnerMainTabViewModel.swift:5]: Thanks, fixed it.

File: Woof/Woof/ViewModel/OwnerMainTabViewModel.swift (hunk 1 above)
Comment (julyakmita) [reply Woof/Woof/ViewModel/OwnerMainTabViewModel.swift:6]: Thanks, fixed it.

Comment (julyakmita) [reply Woof/Woof/ViewModel/OwnerMainTabViewModel.swift:6]: Thanks, for the suggestion. The function was renamed.

Comment (julyakmita) [reply Woof/Woof/ViewModel/OwnerMainTabViewModel.swift:6]: Sorry for this!  The spell checking is enable.

Screenshot 2023-06-13 at 11 21 38 PM

File: Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift (hunk 5 above)
Comment (julyakmita) [reply Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift:1]: My bad! The files are moved.

File: Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift (hunk 6 above)
Comment (julyakmita) [reply Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift:5]: You are right. Thanks for the attention! This test is useless. I removed it from both files.

File: Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift (hunk 8 above)
Comment (julyakmita) [reply Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift:12]: Agree. 
It was renamed on `func testResetCurrentOwnerRoleSuccessfullyChangesUserRoleToNone()`.

[the remark this answers is not in the document]
Comment (julyakmita) [reply Woof/WoofTests/ViewModel/OwnerProfileViewModelTests/SitterMainTabViewModelTests.swift:12]: Fixed.

File: Woof/Woof/ViewModel/OwnerMainTabViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import Foundation
+
+/// The view model for a owner main tab view, responsible for preparing and processing data for the view.
+final class OwnerMainTabViewModel: ObservableObject {
+    /// Resets the current owner role in the app by default value `none`.
```
Comment (NadzeyaShpakouskaya) [inline Woof/Woof/ViewModel/OwnerMainTabViewModel.swift:5]: Does it sound more clear?
```suggestion
    /// Resets the current user role in the app to the default value.
```

File: Woof/Woof/ViewModel/SitterMainTabViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import Foundation
+
+/// The view model for a pet sitter main tab view, responsible for preparing and processing data for the view.
+final class SitterMainTabViewModel: ObservableObject {
+    /// Resets the current sitter role in the app by default value `none`.
```
Comment (NadzeyaShpakouskaya) [inline Woof/Woof/ViewModel/SitterMainTabViewModel.swift:5]: ?
```suggestion
    /// Resets the current user role in the app to the default value.
```

**Review by NadzeyaShpakouskaya — COMMENTED** [review]
Nice improvement, some small suggestion from my side:

File: Woof/Woof/ViewModel/OwnerMainTabViewModel.swift (hunk 11 above)
Comment (julyakmita) [reply Woof/Woof/ViewModel/OwnerMainTabViewModel.swift:5]: Yes, thanks. Fixed it.

File: Woof/Woof/ViewModel/SitterMainTabViewModel.swift (hunk 12 above)
Comment (julyakmita) [reply Woof/Woof/ViewModel/SitterMainTabViewModel.swift:5]: Thanks for the good suggestion. Fixed it.

File: Woof/WoofTests/ViewModel/OwnerMainTabViewModelTests.swift [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+final class OwnerMainTabViewModelTests: XCTestCase {
+    func testResetCurrentRoleMethodExistsInAPI() {
+        OwnerMainTabViewModel().resetCurrentRole()
+    }
+
+    func testResetCurrentRoleSuccessfullyChangesUserRoleToNone() {
+        // Given
+        let initialRole = PreferencesHandler.set(userRole: .owner)
+        XCTAssertNotNil(initialRole)
```
[the remark this answers is not in the document]
Comment (dersim-davaod) [reply Woof/WoofTests/ViewModel/OwnerMainTabViewModelTests.swift:11]: The asserts should be in the `// Then` section.

Please discuss that with teammates and review the rest of code.

File: Woof/WoofTests/ViewModel/OwnerMainTabViewModelTests.swift [hunk 14]
```diff
@@ -0,0 +1,19 @@
+import XCTest
+
+final class OwnerMainTabViewModelTests: XCTestCase {
+    func testResetCurrentRoleMethodExistsInAPI() {
+        OwnerMainTabViewModel().resetCurrentRole()
+    }
+
+    func testResetCurrentRoleSuccessfullyChangesUserRoleToNone() {
```
Comment (dersim-davaod) [inline Woof/WoofTests/ViewModel/OwnerMainTabViewModelTests.swift:8]: nit:

Keep it short:
```suggestion
    func testResetCurrentRoleChangesUserRoleToNone() {
```

**Review by dersim-davaod — CHANGES_REQUESTED** [review]
Consider improvements:

**Review by dersim-davaod — APPROVED** [review]
LGTM

**Review by marinkman — APPROVED** [review]
LGTM

**Review by iBYarchitect — APPROVED** [review]
LGTM

**Review by NadzeyaShpakouskaya — APPROVED** [review]
LGTM

**Pull request merged**